contracts/extension/plugin/Map.sol

```
@@ -3,61 +3,103 @@ pragma solidity ^0.8.0;
+            _setPlugin(_pluginsToRegister[i]);
         }
     }
 
-    function _setExtensions(bytes4 _selector, address _extension) internal {
-        require(allSelectors.add(bytes32(_selector)), "REGISTERED");
+    function setPlugin(Plugin memory _plugin) external {
+        require(_canSetPlugin(), "Not authorized");
 
-        extension[_selector] = _extension;
-        selectorsForExtension[_extension].add(bytes32(_selector));
+        _setPlugin(_plugin);
+    }
+
+    function updatePlugin(Plugin memory _plugin) external {
+        require(_canSetPlugin(), "Not authorized");
+
+        _updatePlugin(_plugin);
+    }
+
+    function removePlugin(bytes4 _selector) external {
+        require(_canSetPlugin(), "Not authorized");
+
+        _removePlugin(_selector);
+    }
+
+    function _setPlugin(Plugin memory _plugin) internal {
+        require(allSelectors.add(bytes32(_plugin.selector)), "REGISTERED");
+        require(_plugin.selector == bytes4(keccak256(abi.encodePacked(_plugin.functionString))), "Incorrect selector");
+
+        pluginForSelector[_plugin.selector] = _plugin;
+        selectorsForPlugin[_plugin.pluginAddress].add(bytes32(_plugin.selector));
 
-        emit ExtensionRegistered(_selector, _extension);
+        emit PluginRegistered(_plugin.selector, _plugin.pluginAddress);
     }
 
-    function getExtensionForFunction(bytes4 _selector) external view returns (address) {
-        address ext = extension[_selector];
-        require(ext != address(0), "No extension available for selector.");
+    function _updatePlugin(Plugin memory _plugin) internal {
+        address currentExtension = getPluginForFunction(_plugin.selector);
+        require(_plugin.selector == bytes4(keccak256(abi.encodePacked(_plugin.functionString))), "Incorrect selector");
 
-        return ext;
+        pluginForSelector[_plugin.selector] = _plugin;
+        selectorsForPlugin[_plugin.pluginAddress].add(bytes32(_plugin.selector));
```

Put this `add` call in a `require` statement.

If `updatePlugin` is called twice with the same `Plugin` struct info, this `add` call will fail silently and the following `remove` call will remove the selector as from the `selectorsForPlugin` set, thus, tainting data / state of the contract.

---

Or, you can reverse the order of the `add` and `remove` calls.